switchapi/es2k/switch_lag.c

```diff
@@ -0,0 +1,472 @@
+#include "switchapi/switch_device.h"
+#include "switchapi/switch_internal.h"
+#include "switchapi/switch_status.h"
+
+/*
+ * Routine Description:
+ *   @brief Create allocated LAG handles
+ *
+ * Arguments:
+ *   @param[in] device - device
+ *
+ * Return Values:
+ *    @return  SWITCH_STATUS_SUCCESS on success
+ *             Failure status code on error
+ */
+switch_status_t switch_lag_init(switch_device_t device) {
+  switch_status_t status = SWITCH_STATUS_SUCCESS;
+
+  status =
+      switch_handle_type_init(device, SWITCH_HANDLE_TYPE_LAG, SWITCH_LAG_MAX);
+  CHECK_RET(status != SWITCH_STATUS_SUCCESS, status);
+
+  status = switch_handle_type_init(device, SWITCH_HANDLE_TYPE_LAG_MEMBER,
+                                   SWITCH_LAG_MAX);
+  CHECK_RET(status != SWITCH_STATUS_SUCCESS, status);
+
+  krnlmon_log_debug("lag init successful on device %d\n", device);
+
+  return status;
+}
+
+/*
+ * Routine Description:
+ *   @brief Free allocated LAG handles
+ *
+ * Arguments:
+ *   @param[in] device - device
+ *
+ * Return Values:
+ *    @return  SWITCH_STATUS_SUCCESS on success
+ *             Failure status code on error
+ */
+switch_status_t switch_lag_free(switch_device_t device) {
+  switch_status_t status = SWITCH_STATUS_SUCCESS;
+
+  status = switch_handle_type_free(device, SWITCH_HANDLE_TYPE_LAG);
+  CHECK_RET(status != SWITCH_STATUS_SUCCESS, status);
+
+  status = switch_handle_type_free(device, SWITCH_HANDLE_TYPE_LAG_MEMBER);
+  CHECK_RET(status != SWITCH_STATUS_SUCCESS, status);
+
+  krnlmon_log_debug("LAG free successful on device %d\n", device);
+
+  return status;
+}
+
+/*
+ * Routine Description:
+ *   @brief Create LAG
+ *
+ * Arguments:
+ *   @param[in] device - device
+ *   @param[in] api_lag_info - LAG info
+ *   @param[out] lag_h - LAG handle
+ *
+ * Return Values:
+ *    @return  SWITCH_STATUS_SUCCESS on success
+ *             Failure status code on error
+ */
+switch_status_t switch_api_lag_create(switch_device_t device,
+                                      switch_api_lag_info_t* api_lag_info,
+                                      switch_handle_t* lag_h) {
+  switch_status_t status = SWITCH_STATUS_SUCCESS;
+  switch_lag_info_t* lag_info = NULL;
+
+  if (api_lag_info->rmac_handle == SWITCH_API_INVALID_HANDLE) {
+    status = switch_api_device_default_rmac_handle_get(
+        device, &api_lag_info->rmac_handle);
+    CHECK_RET(status != SWITCH_STATUS_SUCCESS, status);
+  }
+
+  if (!SWITCH_RMAC_HANDLE(api_lag_info->rmac_handle)) {
+    status = SWITCH_STATUS_INVALID_HANDLE;
+    krnlmon_log_error(
+        "lag create: Invalid rmac handle on device %d: "
+        "error: %s\n",
+        device, switch_error_to_string(status));
+    return status;
+  }
+
+  *lag_h = switch_lag_handle_create(device);
+  CHECK_RET(*lag_h == SWITCH_API_INVALID_HANDLE, SWITCH_STATUS_NO_MEMORY);
+
+  status = switch_lag_get(device, *lag_h, &lag_info);
+  CHECK_RET(status != SWITCH_STATUS_SUCCESS, status);
```

Comment:
lag_h created above at line:113 needs to be deleted before returning here?

Reply:
Addressed.